### PR TITLE
Create merge request when editing on web UI v3 [can be updated]

### PR DESCRIPTION
**Pull request by cirosantilli**
Fixes: ACCEPTING MR at: http://feedback.gitlab.com/forums/176466-general/suggestions/5350102-option-to-make-a-merge-request-when-editing-via-we

Supersedes the original request: https://github.com/gitlabhq/gitlabhq/pull/7266

I proposed custom branch names to GitHub at: https://github.com/isaacs/github/issues/228 and they've half implemented it at: https://github.com/blog/1945-quick-pull-requests 

Dependencies:
- [x] https://github.com/gitlabhq/gitlabhq/pull/8053/files (fixes the only CI failure)
- [ ] https://github.com/gitlabhq/gitlabhq/pull/8257

Dependencies already included here which were extracted from this PR and should be evaluated separately and merged first:
- [ ] https://github.com/gitlabhq/gitlabhq/pull/8128 
- [ ] https://github.com/gitlabhq/gitlabhq/pull/8131
- [ ] https://github.com/gitlabhq/gitlabhq/pull/8183

TODO on future PRs:
- Allow to create new branch without creating a merge request (add new checkbox, implied by create mr)
- When already on fork, add a "Create merge request back to origin checkbox"

TODO just before merging this after having evaluated:
- set the gitlab-shell URL back to the original one
## Blob edit page UI changes

If user has push permission on the current branch, when the page loads it looks like:

screenshot from 2014-11-02 17 12 20 initial

If user clicks on "Create merge request", he now sees:
- the "On my fork" checkbox which shows where the merge request source will be
- the source branch indicator `master` becomes an input so he can choose a custom branch name
- the source repository changes from `gitlab-org` to `root`

screenshot from 2014-11-02 17 13 10 click create mr

By default, branch name is a free branch name on both repositories. When user toggles "On my fork", and he has not yet modified it, it toggles the branch name to a free branch name on given repo:

screenshot from 2014-11-02 17 32 43 on my fork

If user edits the branch name to a custom name like `custom-name`, the branch name toggle stops so as not to lose the user's input. So after editing the branch name and toggle "On my fork", the branch name stays at `custom-name`:

screenshot from 2014-11-02 17 34 40 custom name

User does not have push permission, he must fork. "Create merge request" and "On my fork" are checked and disabled:

screenshot from 2014-10-15 21 23 22 user does not have push permission on current branch

If user is already on his fork, he does not see "On my fork": the merge request will necessarily be generated between the current repo and itself:

screenshot from 2014-11-02 17 15 37 on fork

There is also a minor UI behavior change at blob#show: the edit button is enabled even if you don't have push permission since you can edit to start a merge request:

screenshot from 2014-10-15 21 45 23 blob show edit
## Are there points in the code the reviewer needs to double check?
- The **hard** thing of this MR was the testing. See the dependencies above and check them very well.
- Permissions. Double check the controller specs, and the crux of the implementation: `base_tree_controller.rb`

### CODE REVIEW

**Comment by cirosantilli** [comment]
Github implemented this before us: https://github.com/blog/1945-quick-pull-requests

**Comment by dosire** [comment]
@cirosantilli sorry about that. @randx can you review?

File: app/controllers/projects/base_tree_controller.rb [hunk 1]
```diff
@@ -4,5 +4,116 @@ class Projects::BaseTreeController < Projects::ApplicationController
   before_filter :authorize_read_project!
   before_filter :authorize_download_code!
   before_filter :require_non_empty_project
+
+  protected
+
+  ON_MY_FORK_CHECKBOX_CHECKED_DEFAULT = true
```
Comment (jvanbaarsen) [inline app/controllers/projects/base_tree_controller.rb]: Do you really need this constant here? Isn't there a better way of doing this?

File: app/models/repository.rb [hunk 2]
```diff
@@ -102,6 +106,18 @@ def branch_names
...
+  # Generate a free branch name of form prefix<Number>,
+  # where Number is the smallest unused positive integer.
+  def free_branch_name(prefix)
+    n_taken = []
+    branch_names.each do |name|
+      if name =~ /#{prefix}(\d+)/
+        n_taken << $1.to_i
+      end
+    end
+    "#{prefix}#{((1..n_taken.length + 1).to_a - n_taken)[0]}"
```
Comment (jvanbaarsen) [inline app/models/repository.rb]: Would this also be possible:
```
"#{prefix}#{n_taken.last + 1}"
```

**Comment by jvanbaarsen** [comment]
@cirosantilli Left a few comments. Tip for the future, try and create a bit smaller PR's. Just chop up your feature so refactors are in different PR then the actual feature. That makes reviewing a lot easier.

**Comment by jvanbaarsen** [comment]
This merge request has been closed because a request for more information has not been reacted to for more than 2 weeks. If you respond and conform to the merge request guidelines in our [contributing guidelines](https://gitlab.com/gitlab-org/gitlab-ce/blob/master/CONTRIBUTING.md#pull-requests) we will reopen this merge request.

**Pull request closed without merging**